src/sequencer.ts

```diff
@@ -0,0 +1,305 @@
+import { logger } from './util/logger.js';
+
+export declare type PacketType = 'publish' | 'puback' | 'pubrec' | 'pubrel' | 'pubcomp';
+
+export declare type SequenceType = 'publish';
+
+export interface Message {
+  qos: 0 | 1 | 2;
+  cmd: PacketType;
+  messageId: number | undefined;
+}
+
+export interface Packet {
+  messageId: number | undefined;
+}
+
+// It's called SendPacketFuncton, but it accepts a Message. Maybe we need to rethink this.
+// The more I think about this, the more I think it should accept a `Packet` object, and it
+// should use `Packet.cmd` instead of accepting a separate `PacketType` parameter.
+export type SendPacketFunction = (packetType: PacketType, message: Message) => void;
+
+type InFlightMessageMap = { [key: number]: SequenceMachine };
+
+// These should be documented so callers can change them. Do we want these to be specific for each packet type (pubRelInterval, maxPubRe, etc)?
+/* eslint prefer-const: 0 */
+let pubAckInterval = 2000;
+let maxPublishCount = 5;
+
+type DoneFunction = (err?: Error) => void;
+
+export class MqttPacketSequencer {
+  sendPacketFunction: SendPacketFunction;
+  inFlightMessages: InFlightMessageMap = {};
+
+  constructor(sendPacketFunction: SendPacketFunction) {
+    this.sendPacketFunction = sendPacketFunction;
+  }
+
+  _runSequence(sequenceType: SequenceType, message: Message, done: DoneFunction) {
+    let sequenceMachine: SequenceMachine;
+
+    switch (sequenceType) {
+      case 'publish':
+        switch (message.qos) {
+          case 0:
+            sequenceMachine = new PublishQos0(message, this.sendPacketFunction, done);
+            break;
+          case 1:
+            sequenceMachine = new PublishQos1(message, this.sendPacketFunction, done);
+            break;
+          case 2:
+            sequenceMachine = new PublishQos2(message, this.sendPacketFunction, done);
+            break;
+        }
+        this.inFlightMessages[message.messageId as number] = sequenceMachine;
+      // SUBSCRIBE also goes into the inFlightMesages map. CONNECT maybe goes somewhere else?
+    }
+
+    sequenceMachine.start();
+  }
+
+  // TODO: Is there an easier way to Promisify this?
+  runSequence(sequenceType: SequenceType, message: Message, done?: DoneFunction) {
```

Comment:
could we just get rid of the callback? Supporting both promise and callback patterns can be painful long-term